### ISSUE TITLE
Adding a datepattern in config files is apparently not recognized Fail2Ban v0.10.0a2

### ISSUE THREAD

**Issue by keithchristian** [fail2ban/fail2ban#2352]
### Problem statement:

Using fail2ban as shown in the examples below, the date string is not
recognized in log files (although it is recognized on other machines
using the very same logging format, and using the same version of Python
and Fail2ban....bizarre.)
```
www.xxx.yyy.zzz - - [17/Feb/2019:02:32:05 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 2345 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
```

### Python and Fail2ban version information
[root@someserver ~]# /usr/bin/fail2ban-python -V
Python 2.7.13

[root@someserver ~]# fail2ban-client --version
Fail2Ban v0.10.0a2

### An incorrectly formatted date template string reveals a possible bug 
```
echo;date;fail2ban-regex -d "[%d/%b/%Y:%H:%M:%S +0000]" --print-all-matched /one/fine/directory/python_requests_test_data /etc/fail2ban/filter.d/apache-block-python.conf|head -40

Thu Feb 21 14:31:12 UTC 2019
Traceback (most recent call last):
  File "/usr/bin/fail2ban-regex", line 34, in <module>

Running tests
=============

Use      datepattern : [Day/MON/Year:24hour:Minute:Second +0000]
Use   failregex filter file : apache-block-python, basedir: /etc/fail2ban
Use         log file : /one/fine/directory/python_requests_test_data
Use         encoding : UTF-8

    exec_command_line()
  File "/opt/localpython/lib/python2.7/site-packages/fail2ban/client/fail2banregex.py", line 598, in exec_command_line
    if not fail2banRegex.start(opts, args):
  File "/opt/localpython/lib/python2.7/site-packages/fail2ban/client/fail2banregex.py", line 548, in start
    self.process(test_lines)
  File "/opt/localpython/lib/python2.7/site-packages/fail2ban/client/fail2banregex.py", line 388, in process
    line_datetimestripped, ret = self.testRegex(line)
  File "/opt/localpython/lib/python2.7/site-packages/fail2ban/client/fail2banregex.py", line 349, in testRegex
    ret = self._filter.processLine(line, date)
  File "/opt/localpython/lib/python2.7/site-packages/fail2ban/server/filter.py", line 467, in processLine
    (timeMatch, template) = self.dateDetector.matchTime(l)
  File "/opt/localpython/lib/python2.7/site-packages/fail2ban/server/datedetector.py", line 370, in matchTime
    match = template.matchDate(line)
  File "/opt/localpython/lib/python2.7/site-packages/fail2ban/server/datetemplate.py", line 153, in matchDate
    self._compileRegex()
  File "/opt/localpython/lib/python2.7/site-packages/fail2ban/server/datetemplate.py", line 147, in _compileRegex
    raise e
sre_constants.error: unbalanced parenthesis
```

### Demonstrate that the default date stamp detection does not work on this particular installation                                                 

However, using   -d "%d/%b/%Y:%H:%M:%S" on the command line WORKS

(cat /etc/fail2ban/filter.d/apache-block-python.conf)
```ini
[INCLUDES]

# Read common prefixes. If any customizations available -- read them from
# apache-common.local
before = apache-common.conf

# working date template, unfortunately, adding 'datepattern = %d/%b/%Y:%H:%M:%S' to any config files DOES NOT WORK,
# must use the string below ON THE F2B COMMAND LINE.
#
#       -d "%d/%b/%Y:%H:%M:%S"


[Definition]


#failregex = ^<HOST> .* "POST /servlet/Login HTTP/1.1" .* "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server"
#failregex = ^<HOST> .* tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server"
#failregex = ^<HOST> - - "POST /servlet/Login HTTP/1.1" 200 .* "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server"
#failregex = ^<HOST> - - .* "POST /servlet/Login HTTP/1.1" [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server".*$


#failregex = ^<HOST> .* "POST /servlet/Login HTTP/1.1" 200 923 0 25649 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server"
#failregex = ^<HOST> .* "POST /servlet/Login HTTP/1.1" 200
#failregex = ^<HOST> .* "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server"
#failregex = ^<HOST> .* "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" .*$
#failregex = ^<HOST> .*$
#failregex = <HOST> .* "POST /servlet/Login HTTP/1.1" 200
#failregex = ^<HOST> .* 200
#failregex = ^<HOST> *

failregex = ^<HOST> - - .* "POST /servlet/Login HTTP/1.1" [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server".*$
# missed some failregex = ^<HOST> - - .* "POST /servlet/Login HTTP/1.1" [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server"

ignoreregex =
```

```
[root@someserver ~]# fail2ban-regex --print-all-matched /one/fine/directory/python_requests_test_data /etc/fail2ban/filter.d/apache\-python\-request.conf|cat -n|sed -n '1,50p'
     1
     2  Running tests
     3  =============
     4
     5  Use   failregex filter file : apache-block-python, basedir: /etc/fail2ban
     6  Use      datepattern : Default Detectors
     7  Use         log file : /one/fine/directory/python_requests_test_data
     8  Use         encoding : UTF-8
     9
    10
    11  Results
    12  =======
    13
    14  Failregex: 0 total
    15
    16  Ignoreregex: 0 total
    17
    18  Date template hits:
    19
    20  Lines: 51995 lines, 0 ignored, 0 matched, 51995 missed
    21  [processed in 8.45 sec]
    22
    23  Missed line(s): too many to print.  Use --print-all-missed to print all 51995 lines
```

### Contents of the filter referenced on the command line
(cat /etc/fail2ban/filter.d/apache-block-python.conf)
```ini
[INCLUDES]

# Read common prefixes. If any customizations available -- read them from
# apache-common.local
before = apache-common.conf

# working date template, unfortunately, adding 'datepattern = %d/%b/%Y:%H:%M:%S' to any config files DOES NOT WORK,
# must use the string below ON THE F2B COMMAND LINE.
#
#       -d "%d/%b/%Y:%H:%M:%S"


[Definition]

failregex = ^<HOST> - - .* "POST /servlet/Login HTTP/1.1" [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server".*$
# missed some failregex = ^<HOST> - - .* "POST /servlet/Login HTTP/1.1" [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server"

ignoreregex =
```
### Correct date template string on the command line runs properly  

Which config file should the 'datepattern' template be placed in?
It did not work in the init file, the jail.local file, or the apache-block-python.conf file                                     
```
fail2ban-regex -d "%d/%b/%Y:%H:%M:%S" --print-all-matched /one/fine/directory/python_requests_test_data /etc/fail2ban/filter.d/apache-block-python.conf -n '1,100p'
 
Running tests 
============= 
 
Use      datepattern : Day/MON/Year:24hour:Minute:Second 
Use   failregex filter file : apache-block-python, basedir: /etc/fail2ban 
Use         log file : /one/fine/directory/python_requests_test_data 
Use         encoding : UTF-8 
 
 
Results 
======= 
 
Failregex: 49904 total 
|-  #) [# of hits] regular expression 
|   1) [49904] ^<HOST> - - .* "POST /servlet/Login HTTP/1.1" [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server".*$ 
`- 
 
Ignoreregex: 0 total 
 
Date template hits: 
|- [# of hits] date format 
|  [51995] Day/MON/Year:24hour:Minute:Second 
`- 
 
Lines: 51995 lines, 0 ignored, 49904 matched, 2091 missed 
[processed in 7.21 sec] 
 
|- Matched line(s): 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:04 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 25649 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:04 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 1827 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:04 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 25471 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:04 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 25701 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:04 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 1976 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:04 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 2552 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:04 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 2508 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:04 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 25532 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:04 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 25788 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:05 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 2107 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:05 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 2462 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:05 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 2123 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
|  www.xxx.yyy.zzz - - [17/Feb/2019:02:32:05 +0000] "POST /servlet/Login HTTP/1.1" 200 923 0 2345 "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server" 
```

**Comment by sebres** [comment]
> sre_constants.error: unbalanced parenthesis

1. You has an error in your custom `datepattern`. Note `datepattern` is quasi a regex (with internal substitution of known % date-tokens inside). So you should escape brackets and quantifier operand (`+`)
2. For time-zone better use tag `%z`.

This means you should rewrite it to:
```dif
- -d "[%d/%b/%Y:%H:%M:%S +0000]"
+ -d "\[%d/%b/%Y:%H:%M:%S %z\]"
```

> Which config file should the 'datepattern' template be placed in?

If you try to search here for similar issues (e. g. just type datepattern) you'd very simply find an answer.
For example https://github.com/fail2ban/fail2ban/issues/2330#issuecomment-459965597

Normally you'd find this in manpage [jail.conf.5 - JAIL CONFIGURATION FILE(S) (jail.conf)](https://manpages.debian.org/testing/fail2ban/jail.conf.5.en.html#JAIL_CONFIGURATION_FILE(S)_(_jail.conf)). 
So see there for details.

>`failregex = ^<HOST> - - .* "POST /servlet/Login HTTP/1.1" [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* [0-9][0-9]* "-" "tl3cwl5r3o98/x.y.z kwcasw4wq097/x.y.zq Windows/2021Server".*$`

Your regex may work but is vulnerable (due to both catch-all's `.*`). As well as the last expression `.*$` is to neglect (can be removed without problem, because does nothing) so it is not anchored (neither at start nor at end).
See [#2332 (comment)](https://github.com/fail2ban/fail2ban/issues/2332#issuecomment-460985248) and below for more info why and how it is affected.

**Comment by keithchristian** [comment]
Sebres, thank you for the answer.
I had to use the command line argument so that fail2ban would parse dates and the apache log files at all.
"datepattern =" was not recognized in jail.local, fail2ban.conf, or in apache-block-python.conf.
Does v0.10.0a2 truly support putting it in a file?

**Comment by sebres** [comment]
> Does v0.10.0a2 truly support putting it in a file?

I don't know what some alpha-version does really support (and 0.10.0 was released in 2017, last 10th release is 0.10.4).
In last release it is definitely possible to specify datepattern in filter as well as in the jail configs.

**Comment by sebres** [comment]
Anyway specifying of datepattern in filter-config was always possible already in 0.9, see for example - https://github.com/fail2ban/fail2ban/blob/f2e55e8499744e6b1f9196d2033e25c53c23c61d/config/filter.d/squirrelmail.conf#L7-L9
